### PR TITLE
New Feature saving state as url_fragment

### PR DESCRIPTION
**Pull request by viswaraavi**
The working example for the url_fragment_identifier is 
http://bl.ocks.org/viswaraavi/raw/ddc938aa4b74e21897686b64d195dc12/#sex_smoker={Female.No}

I thinks we need to change the url later to have {} because github is coverting to safe url

### CODE REVIEW

**Comment by viswaraavi** [comment]
The build is passing. I think there is some problem with wrecker build

**Comment by tdhock** [comment]
Hey @viswaraavi thanks for the update. The JS code looks OK to me, but it doesn't seem to be working in my browser (firefox on ubuntu). The plot should render with just one legend entry selected, right? My browser renders it with all 4 legend entries selected.

Also can you please add a test case? Try using `remDr$navigate(url)` as in our shiny test https://github.com/tdhock/animint/blob/master/tests/testthat/test-shiny.R#L19

make sure to set the url back to the original address at the end of your test https://github.com/tdhock/animint/blob/master/tests/testthat/test-shiny.R#L133

**Comment by viswaraavi** [comment]
1.I  think first you need to convert the safeurl to the one with {}
2.And then try to do complete page reload using ctrl+R

**Comment by tdhock** [comment]
not sure I understand what you mean by safe url. when I click your blocks link, I get taken to a url which has {} --- `{Female.No}` 
screenshot-url-fragment

**Comment by viswaraavi** [comment]
Can you try adding sexsmoker={Female.No} manually and check once ?, that way its working for me.

On chrome(ubuntu) I can see that url is being converted to http://bl.ocks.org/viswaraavi/raw/ddc938aa4b74e21897686b64d195dc12/#sex_smoker=%7BFemale.No%7D. I think the same thing is happening in firefox but still firefox is showing the url as normal

**Comment by viswaraavi** [comment]
I am not able to run individual tests of renderer in testthat folder. I have executed all the commands as in debugging workflow2. The sessioninfo gave me this
R version 3.2.3 (2015-12-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04 LTS

  
attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] RSelenium_1.4.0    XML_3.98-1.4       RJSONIO_1.3-0      RCurl_1.95-4.8     bitops_1.0-6       animint_2016.05.11
[7] proto_0.3-10       ggplot2_1.0.1      testthat_1.0.2

**Comment by tdhock** [comment]
still not working on firefox, even after manually typing `sexsmoker` and `sex_smoker` and then reloading

I tried chromium on ubuntu and it did not even render at all, see screenshot
screenshot-chromium-animint-fragment-fails

what browser versions are you using?

```
thocking@silene:~/projects/PeakSegFPOP-paper(master*)$ ~/firefox/firefox --version
Mozilla Firefox 47.0
thocking@silene:~/projects/PeakSegFPOP-paper(master*)$ chromium-browser --version
Chromium 37.0.2062.120 Ubuntu 12.04
thocking@silene:~/projects/PeakSegFPOP-paper(master*)$ 

```

**Comment by tdhock** [comment]
by the way if you want to change the syntax, go ahead. for example instead of `sex_smoker={Female.No}` you could write `sex_smoker___Female.No

**Comment by tdhock** [comment]
about the tests, does the firefox window appear after you execute `tests_init("firefox")`? It should! If not, you probably have some version conflict.

Try one of the version combinations that I listed in the comments in https://github.com/tdhock/animint/blob/master/tests/testthat.R

**Comment by viswaraavi** [comment]
I downgraded my firefox to version 41. The test is running locally. Its working in my browser. I will try ti use __ and try again
image

**Comment by viswaraavi** [comment]
I got the error. I created the gist with previous source where i was using new way to define function in javascript I will create new gist again

**Comment by viswaraavi** [comment]
The test running locally
image

**Comment by viswaraavi** [comment]
http://bl.ocks.org/viswaraavi/raw/314e02462ab0acd4dc7af1b5b85f9aff/#Species={setosa}
I think this should work

**Comment by viswaraavi** [comment]
I have changed the url to look like following
http://bl.ocks.org/viswaraavi/raw/f2bafb827bda3438ac73d60ba19dd425/#Species=setosa

**Comment by viswaraavi** [comment]
The test is running locally can you just check once on your computer.I could not understand why it is nor running with phantomjs

**Comment by tdhock** [comment]
the new format looks good for me, and works on my computer using firefox

your new test is working on my machine using firefox. I also got the following test to work on both firefox and phantomjs:

```r
acontext("url fragment identifier")

iris$observation <- 1:nrow(iris)
viz <- list(
  petal=ggplot()+
    geom_point(aes(Petal.Width, Petal.Length, fill=Species,
                   id=paste0("petal", observation),
                   clickSelects=observation), data=iris),
  sepal=ggplot()+
    geom_point(aes(Sepal.Width, Sepal.Length, fill=Species,
                   id=paste0("sepal", observation),
                   clickSelects=observation), data=iris),
  first=list(observation=5))
info <- animint2HTML(viz)
remDr$screenshot(file="screenshot-url-with-no-selection.png")

test_that("all species are selected for url with no selection", {
  opacity.str <- getStyleValue(info$html, "//td[@class='legend_entry_label']", "opacity")
  opacity.num <- as.numeric(opacity.str)
  opacity.exp <- rep(1, 6)
  expect_equal(opacity.num, opacity.exp)
})

test_that("one observation selected for url with no selection", {
  circle.xpath <- "//circle[@class='geom']"
  node.list <- getNodeSet(info$html, circle.xpath)
  opacity.str <- getStyleValue(info$html, circle.xpath, "opacity")
  names(opacity.str) <- sapply(node.list, xmlAttrs)["id",]
  selected.names <- paste0(c("petal", "sepal"), 5)
  is.selected <- names(opacity.str) %in% selected.names
  expect_equal(as.numeric(opacity.str[is.selected]), rep(1, 2))
  expect_equal(as.numeric(opacity.str[!is.selected]), rep(0.5, 298))
})

old_address <- remDr$getCurrentUrl()[[1]]
new_address <- paste0(old_address, '#Species=setosa')
remDr$navigate(new_address)
remDr$refresh()
Sys.sleep(10)
html <- getHTML()
remDr$screenshot(file="screenshot-url-with-selection.png")

test_that("one species is selected for url with selection", {
  entry.xpath <- "//td[@class='legend_entry_label']"
  node.list <- getNodeSet(html, entry.xpath)
  opacity.str <- getStyleValue(html, entry.xpath, "opacity")
  opacity.num <- as.numeric(opacity.str)
  value.vec <- sapply(node.list, xmlValue)
  is.setosa <- value.vec == "setosa"
  expect_equal(opacity.num[is.setosa], rep(1, 2))
  expect_equal(opacity.num[!is.setosa], rep(0.5, 4))
})

remDr$navigate(old_address)
```

does it work for you using phantomjs?

also I have two questions about your new syntax.

1. how to indicate a selection for more than one variable? say I want to specify a value for the `observation` and `Species` variables.

2. what about selected values that have commas in them?

**Comment by viswaraavi** [comment]
For multiple variable one can specify in following way by using &
http://bl.ocks.org/viswaraavi/raw/f2bafb827bda3438ac73d60ba19dd425/#Species=setosa&id=1

If the variable value have , in them(which is very rare) the  code right now does not work. I will try to put escape character \  to include , in  variable values

**Comment by tdhock** [comment]
also there should be some way for the user to copy a URL with the currently displayed selection state. maybe add that above the Show Selection Menus table? and update it every time `update_selector` is called in the JS code

**Comment by viswaraavi** [comment]
Yeah I was also thinking of it. I will do it

**Comment by tdhock** [comment]
that sounds reasonable about the `&`

about the commas I agree with you it is quite rare, so I think it is fine if we do not support that. In that case we should just have the compiler warn the user that they should remove the `,` and `&` values if they want to use the selection URLs

**Comment by viswaraavi** [comment]
No the tests are not passing with phantomjs I get the following error while running 
`old_address <- remDr$getCurrentUrl()[[1]]`
`[ERROR - 2016-08-12T14:27:28.324Z] RouterReqHand - _handle.error - {"line":40,"stack":"get\n_getUrlCommand\n_handle\n_reroute\n_handle"}`

The tests are passing with firefox

**Comment by viswaraavi** [comment]
I think i will try to support , and & even though , is rare I think & might be used often. So I think it would be better to support them

**Comment by tdhock** [comment]
what version of phantomjs are you using? we use 1.9.8 for testing

```
thocking@silene:~$ R --no-save < test-url-fragment.R 

R version 3.2.3 (2015-12-10) -- "Wooden Christmas-Tree"
Copyright (C) 2015 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library("testthat")
> library("animint")
Loading required package: ggplot2
> library("RSelenium")
Loading required package: RCurl
Loading required package: RJSONIO
Loading required package: XML
> 
> setwd("~/R/animint-viswaraavi/tests/testthat")
> source("helper-functions.R")
> tests_init("phantomjs")
Starting phantomjs binary. To shut it down, run: 
 pJS$stop()
PhantomJS is launching GhostDriver...
[INFO  - 2016-08-12T14:39:31.355Z] GhostDriver - Main - running on port 4444
serving the directory /home/thocking/R/animint-viswaraavi/tests/testthat at http://127.0.0.1:4848
[INFO  - 2016-08-12T14:39:45.339Z] Session [9ccdec70-609a-11e6-8618-21ea4d9e1a2f] - page.settings - {"XSSAuditingEnabled":false,"javascriptCanCloseWindows":true,"javascriptCanOpenWindows":true,"javascriptEnabled":true,"loadImages":true,"localToRemoteUrlAccessEnabled":false,"userAgent":"Mozilla/5.0 (Unknown; Linux x86_64) AppleWebKit/534.34 (KHTML, like Gecko) PhantomJS/1.9.8 Safari/534.34","webSecurityEnabled":true}
[INFO  - 2016-08-12T14:39:45.339Z] Session [9ccdec70-609a-11e6-8618-21ea4d9e1a2f] - page.customHeaders:  - {}
[INFO  - 2016-08-12T14:39:45.339Z] Session [9ccdec70-609a-11e6-8618-21ea4d9e1a2f] - Session.negotiatedCapabilities - {"browserName":"phantomjs","version":"1.9.8","driverName":"ghostdriver","driverVersion":"1.1.0","platform":"linux-unknown-64bit","javascriptEnabled":true,"takesScreenshot":true,"handlesAlerts":false,"databaseEnabled":false,"locationContextEnabled":false,"applicationCacheEnabled":false,"browserConnectionEnabled":false,"cssSelectorsEnabled":true,"webStorageEnabled":false,"rotatable":false,"acceptSslCerts":false,"nativeEvents":true,"proxy":{"proxyType":"direct"}}
[INFO  - 2016-08-12T14:39:45.340Z] SessionManagerReqHand - _postNewSessionCommand - New Session Created: 9ccdec70-609a-11e6-8618-21ea4d9e1a2f
> 
> iris$observation <- 1:nrow(iris)
> viz <- list(
+   petal=ggplot()+
+     geom_point(aes(Petal.Width, Petal.Length, fill=Species,
+                    id=paste0("petal", observation),
+                    clickSelects=observation), data=iris),
+   sepal=ggplot()+
+     geom_point(aes(Sepal.Width, Sepal.Length, fill=Species,
+                    id=paste0("sepal", observation),
+                    clickSelects=observation), data=iris),
+   first=list(observation=5))
> info <- animint2HTML(viz)
> remDr$screenshot(file="screenshot-url-with-no-selection.png")
> 
> test_that("all species are selected for url with no selection", {
+   opacity.str <- getStyleValue(info$html, "//td[@class='legend_entry_label']", "opacity")
+   opacity.num <- as.numeric(opacity.str)
+   opacity.exp <- rep(1, 6)
+   expect_equal(opacity.num, opacity.exp)
+ })
> 
> test_that("one observation selected for url with no selection", {
+   circle.xpath <- "//circle[@class='geom']"
+   node.list <- getNodeSet(info$html, circle.xpath)
+   opacity.str <- getStyleValue(info$html, circle.xpath, "opacity")
+   names(opacity.str) <- sapply(node.list, xmlAttrs)["id",]
+   selected.names <- paste0(c("petal", "sepal"), 5)
+   is.selected <- names(opacity.str) %in% selected.names
+   expect_equal(as.numeric(opacity.str[is.selected]), rep(1, 2))
+   expect_equal(as.numeric(opacity.str[!is.selected]), rep(0.5, 298))
+ })
> 
> old_address <- remDr$getCurrentUrl()[[1]]
> new_address <- paste0(old_address, '#Species=setosa')
> remDr$navigate(new_address)
> remDr$refresh()
> Sys.sleep(10)
> html <- getHTML()
> remDr$screenshot(file="screenshot-url-with-selection.png")
> 
> test_that("one species is selected for url with selection", {
+   entry.xpath <- "//td[@class='legend_entry_label']"
+   node.list <- getNodeSet(html, entry.xpath)
+   opacity.str <- getStyleValue(html, entry.xpath, "opacity")
+   opacity.num <- as.numeric(opacity.str)
+   value.vec <- sapply(node.list, xmlValue)
+   is.setosa <- value.vec == "setosa"
+   expect_equal(opacity.num[is.setosa], rep(1, 2))
+   expect_equal(opacity.num[!is.setosa], rep(0.5, 4))
+ })
> 
> remDr$navigate(old_address)
> 
> devtools::session_info()
Session info -------------------------------------------------------------------
 setting  value                       
 version  R version 3.2.3 (2015-12-10)
 system   x86_64, linux-gnu           
 ui       X11                         
 language en_CA:en                    
 collate  en_CA.UTF-8                 
 tz       <NA>                        
 date     2016-08-12                  

Packages -----------------------------------------------------------------------
 package    * version    date       source                             
 animint    * 2016.07.12 2016-08-12 local                              
 bitops       1.0-6      2013-08-17 CRAN (R 3.0.2)                     
 caTools      1.17.1     2014-09-10 CRAN (R 3.2.3)                     
 colorspace   1.2-6      2015-03-11 CRAN (R 3.2.3)                     
 crayon       1.3.1      2015-07-13 CRAN (R 3.2.3)                     
 devtools     1.11.1     2016-04-21 CRAN (R 3.2.3)                     
 digest       0.6.10     2016-08-02 CRAN (R 3.2.3)                     
 ggplot2    * 2.1.0      2016-03-01 CRAN (R 3.2.3)                     
 gtable       0.2.0      2016-02-26 CRAN (R 3.2.3)                     
 labeling     0.3        2014-08-23 CRAN (R 3.2.3)                     
 memoise      1.0.0      2016-01-29 CRAN (R 3.2.3)                     
 munsell      0.4.3      2016-02-13 CRAN (R 3.2.3)                     
 plyr         1.8.4      2016-06-08 CRAN (R 3.2.3)                     
 Rcpp         0.12.6     2016-07-19 CRAN (R 3.2.3)                     
 RCurl      * 1.96-0     2016-08-03 local                              
 RJSONIO    * 1.3-1      2016-08-03 local                              
 RSelenium  * 1.3.6      2016-08-11 Github (ropensci/RSelenium@f597355)
 scales       0.4.0      2016-02-26 CRAN (R 3.2.3)                     
 testthat   * 0.11.0     2015-10-14 CRAN (R 3.2.3)                     
 withr        1.0.1      2016-02-04 CRAN (R 3.2.3)                     
 XML        * 3.99-0     2016-08-03 local                              
> system("phantomjs --version")
1.9.8
> 
thocking@silene:~$ 
```

**Comment by tdhock** [comment]
the code in that script is https://gist.github.com/tdhock/44ef5479f65636c0103d708c1faa35da

the output screenshots are

first one with no selector url

screenshot-url-with-no-selection

second one with selector url

screenshot-url-with-selection

**Comment by viswaraavi** [comment]
I am using 2.1.1. I will try to downgrade and see

**Comment by viswaraavi** [comment]
I have changed the way the url fragment is used to previous version. The previous problem can be easily eliminated by decodeurl function. So there will be no need to worry of , &. 
The example for this is
http://bl.ocks.org/viswaraavi/raw/0b10353996ac51fba682cd91594acfa1/

**Comment by tdhock** [comment]
great work, thanks for the update!

Nice to hear about the support for `&` and `,` -- make sure to add tests for selected values with those characters.

did you ever get the tests working with phantomjs?

File: tests/testthat/test-renderer1-url-fragment.R [hunk 1]
```diff
@@ -1,26 +1,55 @@
...
+                   id=paste0("petal", observation),
+                   clickSelects=observation), data=iris),
+  sepal=ggplot()+
+    geom_point(aes(Sepal.Width, Sepal.Length, fill=Species,
+                   id=paste0("sepal", observation),
+                   clickSelects=observation), data=iris),
+  first=list(observation=5))
 
+info <- animint2HTML(viz)
+remDr$screenshot(file="screenshot-url-with-no-selection.png")
```
Comment (tdhock) [inline tests/testthat/test-renderer1-url-fragment.R]: please do not take screenshots in the real test file

**Comment by viswaraavi** [comment]
Yeah I got the tests working for downgraded version

**Comment by tdhock** [comment]
great that you got the tests working!

also the Current URL display is nice, but could you please move it to a row of the "Show selection menus" table? that way the user can find it if they want it, but it won't take up so much space below the data viz.

**Comment by tdhock** [comment]
and please add a test that checks for updates to the Current URL row, after changing the selection via a mouse click.

**Comment by tdhock** [comment]
actually now that I think of it, is it possible to have the JavaScript code change the URL that is displayed in the browser address bar? That would be even better than the Current URL row in the web page content.

**Comment by viswaraavi** [comment]
It's possible http://stackoverflow.com/questions/824349/modify-the-url-without-reloading-the-page. But when we push the state to that url the animation does not start immediately because I have coded in such a way that when there is url fragment the animation to stop. And the user can only start animation on clicking play. in animation controls. So the default feature where animation start's loading on the page will change.

**Comment by tdhock** [comment]
hm. that seems reasonable to not start the animation, when a selection is specified in the URL.

however I think it is important that the animation does start automatically, when NO selection is specified in the URL.

Is it possible to have both?

**Comment by viswaraavi** [comment]
I dont think so there is any way i can differentiate the both within code

**Comment by viswaraavi** [comment]
Example for the change
http://bl.ocks.org/viswaraavi/raw/95f2c2b6a5d03d63e2ef07d26950d7af/

**Comment by tdhock** [comment]
that looks great!

**Comment by viswaraavi** [comment]
I think if we need to use browser's url for showing current url then may be if we pass extra parameter like 
animation=true to the url and user can change the parameters according to his needs

**Comment by tdhock** [comment]
that sounds reasonable, but would we be able to infer animation=true even if nothing is specified in the URL? (but animation is specified in the JSON)

**Comment by tdhock** [comment]
anyway it is also totally reasonable to just have the link under Show selection menus

**Comment by tdhock** [comment]
also, did you know that it is possible to have more than one animint per web page? for example http://cbio.mines-paristech.fr/~thocking/animint-book/Ch02-ggplot2.html

I don't think it is necessary to support that, but just thought I would mention it.

**Comment by viswaraavi** [comment]
What I am thinking to do is 
By default animation true would be true so the animation starts immediately
If the user wants to go to particular selector url then he has to make himself animation to be false so that the animation would stop then.

The reason why I am thinking of browser urls is because we get extra functionality like going back(which is like undo) and moving forward( which is like redo)

**Comment by viswaraavi** [comment]
I think to support for multiple plots then we can't use browser url we have to go for what we are doing

**Comment by tdhock** [comment]
interesting

can you also update the onscreen URL with animationTRUE or animationFALSE, after each time somebody clicks on play/pause?

like i said no problem about not supporting multiple plots. its already amazing if we can get it to work for single plots

**Comment by viswaraavi** [comment]
http://bl.ocks.org/viswaraavi/raw/1d5f78bdc2a041570bee2f6cb1dbbd63/#Species={setosa,virginica,versicolor}id={1}

There is actually no need of animation variable it workcs because pushstate does not reload the page.
The back functionality works but it only goes back once. And also once he goes back the animation stops 
playing and to support the back functionality i think we need to pass the animation variable

**Comment by tdhock** [comment]
wow, that's great that you got the URL in the browser URL bar working. Thanks!

I'd like to poll the other animint devs to ask which version of this feature they would prefer. @cpsievert @caijun @srvanderplas @kferris10 @faizan-khan-iit any opinion about which of the two animints below you would prefer?

1. every time the selection state changes, an URL on the web page is updated http://bl.ocks.org/viswaraavi/raw/0b10353996ac51fba682cd91594acfa1/
2. every time the selection state changes, the URL in the browser URL bar changes http://bl.ocks.org/viswaraavi/raw/1d5f78bdc2a041570bee2f6cb1dbbd63/

originally I thought option 2 would be more user-friendly, but now I am thinking that I prefer option 1. It is actually kind of annoying to me that every time I change the animint state, it adds one more page to the browser history.

**Comment by caijun** [comment]
Wow, it is a great feature. I prefer option 1 as it save space and users get used to copy the URL in the browser URL bar to share a webpage.

**Comment by cpsievert** [comment]
Yea, I like option 1b as well, nice work @viswaraavi!

**Comment by faizan-khan-iit** [comment]
@viswaraavi This looks awesome! Great work!
@tdhock I would prefer 1b too. More practical and looks neater.

**Comment by viswaraavi** [comment]
So I would start writing tests for 1.b

**Comment by tdhock** [comment]
great! when you think everything is ready to merge, make sure to add some inline comments that explain your code and tests. I will be on vacation for the next two weeks so if I am slow to respond, then maybe you can ask @cpsievert or @faizan-khan-iit to review and merge your code.

**Comment by tdhock** [comment]
any updates?

**Comment by viswaraavi** [comment]
I will try to work on both these pull requests this weekend and try to complete them there is only some small work to be done

File: inst/htmljs/animint.js [hunk 2]
```diff
@@ -1781,12 +1781,17 @@ var animint = function (to_select, json_file) {
       return selected_values;
   };
   
-    
+  var counter=-1;
```
Comment (viswaraavi) [inline inst/htmljs/animint.js]: Update selector url is a function that is used to update the url at the bottom of selection menu

File: inst/htmljs/animint.js [hunk 3]
```diff
@@ -6,10 +6,11 @@
 // Constructor for animint Object.
 var animint = function (to_select, json_file) {
 
-  function wait_until_then(timeout, condFun, readyFun) {
+   function wait_until_then(timeout, condFun, readyFun) {
```
Comment (viswaraavi) [inline inst/htmljs/animint.js]: This change is to support functions that need to be run asynchronously and take arguments

**Comment by tdhock** [comment]
hey @viswaraavi sorry it's taken me so long to code review your PR.

Everything seems to be working fine on my computer. The new code looks good, and the tests are working. Thanks for your PR!

**Pull request merged**